addons/hr_payroll/models/hr_leave.py

```diff
@@ -71,3 +129,33 @@ def action_refuse(self):
         benefits = self.env['hr.benefit'].search([('leave_id', 'in', self.ids)])
         benefits.write({'display_warning': False, 'leave_id': None})
         return True
+
+    def _get_number_of_days(self, date_from, date_to, employee_id):
+        """ If an employee is currently working full time but requests a leave next month
+            where he has a new contract working only 3 days/week. This should be taken into
+            account when computing the number of days for the leave (2 weeks leave = 6 days).
+            Override this method to get number of days according to the contract's calendar
+            at the time of the leave.
+        """
+        super(HrLeave, self)._get_number_of_days(date_from, date_to, employee_id)
+        if employee_id:
+            employee = self.env['hr.employee'].browse(employee_id)
+            # Use sudo otherwise base users can't compute number of days
+            contracts = employee.sudo()._get_contracts(date_from, date_to, states=['incoming', 'open', 'pending'])
+            calendar = contracts[:1].resource_calendar_id if contracts else None # Note: if len(contracts)>1, the leave creation will crash because of unicity constaint
+            return employee.get_work_days_data(date_from, date_to, calendar=calendar)['days']
+
+        time_delta = date_to - date_from
+        return math.ceil(time_delta.days + float(time_delta.seconds) / 86400)
+
+    @api.multi
+    @api.depends('number_of_days')
+    def _compute_number_of_hours_display(self):
```

Comment:
Check that the code is still correct see fix (https://github.com/odoo/odoo/commit/765075ff475e0120d1c3f5564f2259c34d2386cc)